x-pack/plugins/task_manager/server/lib/throughput_manager.ts

```diff
@@ -0,0 +1,173 @@
+const POLL_INTERVAL_DECREASE_PERCENTAGE = 0.95;
+const POLL_INTERVAL_INCREASE_PERCENTAGE = 1.2;
+
+interface DynamicConfiguration<T> {
+  startingValue: T;
+  currentValue: T;
+  observable$: BehaviorSubject<T>;
+}
+
+interface ThroughputManagerOpts {
+  maxWorkers$: BehaviorSubject<number>;
+  pollInterval$: BehaviorSubject<number>;
+  startingMaxWorkers: number;
+  startingPollInterval: number;
+  errors$: Observable<Error>;
+  logger: Logger;
+}
+
+export class ThroughputManager {
+  private throughputCheckSubcription?: Subscription;
+  private readonly throughputCheck$: Observable<number>;
+  private readonly logger: Logger;
+  private readonly maxWorkers: DynamicConfiguration<number>;
+  private readonly pollInterval: DynamicConfiguration<number>;
+  private readonly errors$: Observable<Error>;
+
+  constructor(opts: ThroughputManagerOpts) {
+    this.logger = opts.logger;
+    this.errors$ = opts.errors$;
+    this.maxWorkers = {
+      startingValue: opts.startingMaxWorkers,
+      currentValue: opts.startingMaxWorkers,
+      observable$: opts.maxWorkers$,
+    };
+    this.pollInterval = {
+      startingValue: opts.startingPollInterval,
+      currentValue: opts.startingPollInterval,
+      observable$: opts.pollInterval$,
+    };
+    // Count the number of errors from errors$ and reset whenever throughputCheckInterval$ flushes it
+    this.throughputCheck$ = merge(
+      // Flush error count at fixed interval
+      interval(ADJUST_THROUGHPUT_INTERVAL).pipe(map(() => FLUSH_MARKER)),
+      this.errors$.pipe(filter((e) => SavedObjectsErrorHelpers.isTooManyRequestsError(e)))
+    ).pipe(
+      // When tag is "flush", reset the error counter
+      // Otherwise increment the error counter
+      mergeScan(
+        ({ count }, next) =>
+          next === FLUSH_MARKER
+            ? of(emitErrorCount(count), resetErrorCount())
+            : of(incementErrorCount(count)),
+        emitErrorCount(0)
+      ),
+      filter(isEmitEvent),
+      map(({ count }) => count)
+    );
+  }
+
+  public get isStarted() {
+    return this.throughputCheckSubcription && !this.throughputCheckSubcription.closed;
+  }
+
+  public start() {
+    if (!this.isStarted) {
+      this.throughputCheckSubcription = this.throughputCheck$.subscribe((errorCount) => {
+        if (errorCount > 0) {
+          this.reduceThroughput(errorCount);
+          return;
+        }
+        this.increaseThroughput();
+      });
+    }
+  }
+
+  public stop() {
+    this.throughputCheckSubcription?.unsubscribe();
+    // Reset observable values to original values
+    this.updateConfiguration(this.maxWorkers.startingValue, this.pollInterval.startingValue);
+  }
+  private reduceThroughput(errorCount: number) {
+    const newMaxWorkers = Math.max(
+      Math.floor(this.maxWorkers.currentValue * MAX_WORKERS_DECREASE_PERCENTAGE),
+      1
+    );
+    const newPollInterval = Math.ceil(
+      this.pollInterval.currentValue * POLL_INTERVAL_INCREASE_PERCENTAGE
+    );
```

Comment:
It's worth adding details comments explaining the calculation.
Math doesn't come naturally to me and I was sure this was a mistake for a solid minute... forgetting _smaller=faster_ with intervals.